File: backend/app/models/booking.py
from decimal import Decimal
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.db.base import Base


class Booking(Base):
    """Customer booking order. Holds tickets in PENDING_PAYMENT state until paid."""
    __tablename__ = "bookings"

    booking_id = Column(Integer, primary_key=True, index=True)
    schedule_id = Column(Integer, ForeignKey("event_schedules.schedule_id"), nullable=False, index=True)

    # Customer info (no account required)
    customer_name = Column(String(255), nullable=False)
    phone = Column(String(50), nullable=False)
    email = Column(String(255), nullable=False)
    payment_account = Column(Text, nullable=True)

    payment_status = Column(String(50), nullable=True, default="Pending")
    created_at = Column(DateTime, nullable=True)

    schedule = relationship("EventSchedule", back_populates="bookings")
    e_tickets = relationship("ETicket", back_populates="booking")
    payment_transactions = relationship("PaymentTransaction", back_populates="booking")
    refund_transactions = relationship("RefundTransaction", back_populates="booking")

    @property
    def booking_status(self):
        status_map = {
            "Pending": "PENDING_PAYMENT",
            "Paid": "PAID",
            "Failed": "PAYMENT_FAILED",
            "Refunding": "REFUNDING",
            "Refunded": "REFUNDED",
        }
        return status_map.get(self.payment_status, self.payment_status or "PENDING_PAYMENT")

    @booking_status.setter
    def booking_status(self, value):
        status_map = {
            "PENDING_PAYMENT": "Pending",
            "PAID": "Paid",
            "PAYMENT_FAILED": "Failed",
            "REFUNDING": "Refunding",
            "REFUNDED": "Refunded",
        }
        self.payment_status = status_map.get(value, value)

    @property
    def total_amount(self):
        return sum((Decimal(str(ticket.ticket_config.price)) for ticket in self.e_tickets if ticket.ticket_config), Decimal("0"))

    @property
    def expires_at(self):
        return None

    @property
    def updated_at(self):
        return self.created_at

File: frontend/src/types/booking.ts
import type { ETicket } from "./ticket";

export type BookingStatus =
  | "PENDING_PAYMENT"
  | "PAID"
  | "PAYMENT_FAILED"
  | "CANCELLED"
  | "REFUNDING"
  | "REFUNDED"
  | "REFUND_FAILED"
  | string;

export type ReservationRequest = {
  customer_name: string;
  phone: string;
  email: string;
  schedule_id: number;
  event_day_id: number;
  ticket_ids: number[];
};

export type BookingRead = {
  booking_id: number;
  schedule_id: number;
  customer_name: string;
  phone: string;
  email: string;
  payment_account?: string | null;
  booking_status: BookingStatus;
  total_amount: string | number;
  expires_at?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  e_tickets: ETicket[];
};

export type BookingListItem = {
  booking_id: number;
  customer_name: string;
  email: string;
  phone: string;
  booking_status: BookingStatus;
  total_amount: string | number;
  created_at?: string | null;
};
